OrbitSsh/include/OrbitSsh/Task.h

```diff
@@ -0,0 +1,134 @@
+// Copyright (c) 2020 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef ORBIT_SSH_TASKfine ORBIT_SSH_TASK_H_
+
+#include <functional>
+#include <memory>
+#include <optional>
+#include <string>
+
+#include "Channel.h"
+
+namespace OrbitSsh {
+
+// Task manages a ssh exec channel. An exec channel executes one
+// command on the remote server, periodically returns its output and returns the
+// exit code when the command returns. The Tick function should be called
+// periodically until the command finishes to receive the commands output.
+// The class is constructed with the command to be executed and three callbacks.
+// std_out_callback is called whenever the command produces stdout output,
+// std_err_callback whenever it produces stderr output and exit_callback is
+// called once when either the command exits on its own.
+template <typename ResultType>
+class Task {
+ public:
+  Task(
+      Session* session_ptr, std::string command,
+      std::function<void(std::string, std::optional<ResultType>*)>
+          std_out_callback,
+      std::function<void(std::string, std::optional<ResultType>*)>
+          std_err_callback,
+      std::function<ResultType(int, std::optional<ResultType>*)> exit_callback);
+
+  outcome::result<ResultType> Run();
+
+ private:
+  enum class State { kNotInitialized, kChannelOpened, kRunning, kFinished };
+
+  outcome::result<void> Read();
+  std::optional<ResultType> result_ = std::nullopt;
+  Session* session_ptr_;
+  std::optional<Channel> channel_;
+  std::string command_;
+  std::function<void(std::string, std::optional<ResultType>*)>
+      std_out_callback_;
+  std::function<void(std::string, std::optional<ResultType>*)>
+      std_err_callback_;
+  std::function<ResultType(int, std::optional<ResultType>*)> exit_callback_;
```

Comment:
Can you add type aliases with `using` for these std::functions? Then they will get a lot shorter.

Reply:
done